08openmp/index.md

```diff
@@ -0,0 +1,307 @@
+
+The `parallel` directive tells OpenMP that we want to parallelise the piece of code inside the following scope. What this really means is when execution reaches this parallel region, OpenMP will create a number of threads, allow any constructs inside the parallel region to use those threads, and then destroy those same threads. Strictly, the `parallel` construct *does nothing by itself* except create and destroy threads.
+
+The `for` construct in this example uses the threads spawned by `parallel` by splitting the iterations of the for loop into chunks and assigning different chunks to different threads. Default behaviour can change based on compiler and OpenMP runtime but usually the `for` construct will split the loop into as many chunks as their are threads and assign approximately the same number of iterations to each thread. So if there are 16 loop iterations to be performed at 4 threads, you can probably expect one thread to take iterations 0-4, another to take 5-8, and so on. Not all loops are this simple so we'll discuss later how we can optimise the `for` construct by specifying how the iterations are split amongst the threads.
+
+### Example 1: Filling an array
+
+Let's apply this `parallel for` to an example where we want to fill an array with values from a moderately expensive function like `sin`:
+
+```cpp
+#include <iostream>
+#include <vector>
+#include <cmath>
+#include "timer.hpp"
+
+using namespace std;
+
+int main() {
+  const int N = 100'000'000;
+  vector<double> vec(N);
+
+  Timer timer;
+
+#pragma omp parallel for
+  for(int i=0; i<vec.size(); ++i) {
+    vec[i] = sin(M_PI*float(i)/N);
+  }
+
+  std::cout << timer.elapsed() << '\n';
+
+  return 0;
+}
+```
+
+You can find this example and all further examples in the examples repository TODO. The timer code in the example is relatively simple so I won't explain it in detail, it just starts a timer when it's created and shows the elapsed time when `timer.elapsed()` is called.
+
+Let's build this *without* OpenMP to
+
+1. show that the compiler will simply ignore the pragmas if we don't explicitly tell it to use OpenMP and
+2. measure the time of the serial code.
+
+Building this with `g++ main.cpp timer.cpp` and running with `./a.out` prints out the time taken to fill this vector. It should take less than a second or so on your machine. Running on my laptop once, I got a time of $0.98$s. Adding the OpenMP flag for GCC, compiling with `g++ -fopenmp main.cpp timer.cpp` and again running with `./a.out` gives me a time of $0.14$. For me that's a speed up of around $700$% with one additional line of code! Given I have an 8-core processor, $800$% would be ideal but perfect scaling isn't usually possible in parallel computations, and perhaps with some optimisation we could squeeze out a little more performance. What performance increase do you see when running on your machine and does it match what you expect from the number of cores in your CPU?
```

Comment:
This is clearer than the previous example, but "speed up" is still not technically defined, which it probably should be given that it's an important and standard measure.

Reply:
You might want also to describe how they can check the number of cores on their CPU (or link to a source) as there's a good chance they won't know how many cores their CPU has!